### PR TITLE
[wheel] Move build-wheels out of dev

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
This is about 1300 LOC to platform review, but a lot of it is somewhat boilerplate-y.

Curated as two separate commits, to merge using a merge commit.  This correctly preserves git history even in the presence of adding back the symlink stub for dev CI.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16211)

### CODE REVIEW

**Comment by jwnimmer-tri** [comment]
@drake-jenkins-bot linux-focal-unprovisioned-gcc-wheel-experimental-snopt-release please

**Comment by jwnimmer-tri** [comment]
+@mwoehlke-kitware for feature review, please.

+@rpoyner-tri for platform review, please.  Note in particular that this is moving about 1300 LOC from dev into non-dev, so you'll need to expand and review those files as-new, even though they show up in Reviewable by default as "no diffs".

**Comment by jwnimmer-tri** [comment]
\CC @svenevs FYI -- because this is a straight rename, hopefully this merges up easily with your VTK 9 PR #16132 (for whoever merges first).

**Comment by svenevs** [comment]
Rebases cleanly :slightly_smiling_face:

**Comment by mwoehlke-kitware** [comment]
:lgtm: assuming none of the existing code ends up being changed. (Otherwise I'll probably want to look at the changes.)

**Review by rpoyner-tri — COMMENTED** [review]
:lgtm: minor documentation remarks -- feel free to push back or defer to another pass.

> Reviewed 43 of 44 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16211#-:-MqujFV--gVW8sEvqORr:b7e6cb)* status: 2 unresolved discussions, labeled "do not merge", when planning a "properly curated" merge commit the PR must always be rebased onto latest master (waiting on @jwnimmer-tri and @mwoehlke-kitware)

___
*[tools/wheel/build-wheels, line 3 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16211#-MqtvjWJ2Z9uNINQ_rR2:-MqtvjWJ2Z9uNINQ_rR3:b-xhtqr4) ([raw file](https://github.com/robotlocomotion/drake/blob/64fffa09d0db79c5246d0e1fffbea02e7451e401/tools/wheel/build-wheels#L3)):*
> ```
> #!/usr/bin/env python3
> 
> # This script uses Docker to build the PyPI wheels; run with --help for
> ```

BTW this could be a file-scope docstring.  That would open up the possibility of using this text in help messages.
___
*[tools/wheel/build-wheels, line 234 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16211#-MqubRG89dAHfOcI3dOP:-MqubRG89dAHfOcI3dOQ:btw0v16) ([raw file](https://github.com/robotlocomotion/drake/blob/64fffa09d0db79c5246d0e1fffbea02e7451e401/tools/wheel/build-wheels#L234)):*
> ```
> 
> 
> def check_version(version):
> ```

Can we document somehow what set the giant regex is supposed to match? I understand it may difficult or error-prone to render the match into English, or even to list enough examples to illuminate interesting cases. Maybe there is some external spec we can refer to?
___
*[tools/wheel/image/setup.py, line 47 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16211#-MqugZh7ALmrgYrhdryL:-MqugZh7ALmrgYrhdryM:bp0p809) ([raw file](https://github.com/robotlocomotion/drake/blob/64fffa09d0db79c5246d0e1fffbea02e7451e401/tools/wheel/image/setup.py#L47)):*
> ```Python
> Lab (CSAIL). It is a collection of tools for analyzing the dynamics of
> robots and building control systems for them in C++ and Python, with a
> heavy emphasis on optimization-based design/analysis.'''.strip(),
> ```

This long_description seems stale compared to the one currently at drake.mit.edu. Can we at least update it, or better, avoid hard-coding it here?

**Review by jwnimmer-tri — COMMENTED** [review]
@mwoehlke-kitware I'll let you weigh in on the changes here before I merge.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16211#-:-MqunufK8el57Bjm_XFe:bq0x0ij)* status: labeled "do not merge" (waiting on @mwoehlke-kitware and @rpoyner-tri)

___
*[tools/wheel/build-wheels, line 3 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16211#-MqtvjWJ2Z9uNINQ_rR2:-Mqulo-b9KUNaVdWoysz:b-siz3ls) ([raw file](https://github.com/robotlocomotion/drake/blob/64fffa09d0db79c5246d0e1fffbea02e7451e401/tools/wheel/build-wheels#L3)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

BTW this could be a file-scope docstring.  That would open up the possibility of using this text in help messages.
</blockquote></details>

OK I don't think that's a meaningful improvement.
___
*[tools/wheel/build-wheels, line 234 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16211#-MqubRG89dAHfOcI3dOP:-Mqunssu4IwjDHTYcusq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/64fffa09d0db79c5246d0e1fffbea02e7451e401/tools/wheel/build-wheels#L234)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

Can we document somehow what set the giant regex is supposed to match? I understand it may difficult or error-prone to render the match into English, or even to list enough examples to illuminate interesting cases. Maybe there is some external spec we can refer to?
</blockquote></details>

Done.
___
*[tools/wheel/image/setup.py, line 47 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16211#-MqugZh7ALmrgYrhdryL:-MqunRzv2DBlDjor0C72:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/64fffa09d0db79c5246d0e1fffbea02e7451e401/tools/wheel/image/setup.py#L47)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

This long_description seems stale compared to the one currently at drake.mit.edu. Can we at least update it, or better, avoid hard-coding it here?
</blockquote></details>

Done.

**Comment by jwnimmer-tri** [comment]
@drake-jenkins-bot linux-focal-unprovisioned-gcc-wheel-experimental-snopt-release please

**Review by rpoyner-tri — COMMENTED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16211)* status: 1 unresolved discussion, labeled "do not merge" (waiting on @jwnimmer-tri and @mwoehlke-kitware)

___
*[tools/wheel/build-wheels, line 235 at r3](https://reviewable.io/reviews/robotlocomotion/drake/16211#-MqupT9jAacnFIZOdV7d:-MqupT9jAacnFIZOdV7e:b-pskqsc) ([raw file](https://github.com/robotlocomotion/drake/blob/b0e08d844c5b58ef8d853ee1d137639a86d17753/tools/wheel/build-wheels#L235)):*
> ```
> mathces
> ```
typo

File: tools/wheel/build-wheels
Comment (mwoehlke-kitware) [inline tools/wheel/build-wheels:235]: > Can we document somehow what set the giant regex is supposed to match? I understand it may difficult or error-prone to render the match into English, or even to list enough examples to illuminate interesting cases. Maybe there is some external spec we can refer to?

:+1:, this looks good. Probably should have been in the original version; sorry!

Note that the regex itself is almost verbatim from [PEP 440 Appendix B](https://www.python.org/dev/peps/pep-0440/#appendix-b-parsing-version-strings-with-regular-expressions). I split it into multiple parts for line length and adapted/simplified the additional bit to match the local version segment, which for some reason is missing from the sample `is_canonical`.

**Review by mwoehlke-kitware — APPROVED** [review]
LGTM.

Reviewable is not working again:
>  Failed to publish: Request queued but server did not respond

**Review by jwnimmer-tri — COMMENTED** [review]
> Dismissed @mwoehlke-kitware from a discussion.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16211)* status: :shipit: complete! all discussions resolved, LGTM from assignees rpoyner-tri(platform),mwoehlke-kitware, base commit is latest master (waiting on @jwnimmer-tri)

**Comment by jwnimmer-tri** [comment]
Enough of the builds have passed that I'm going to force-merge this now, to avoid rebasing races.

**Pull request merged**
